File: DataStructures/StacksAndQueues/Queues/MaxPQ.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* Function Prototypes */
int insert(int item);
int is_empty();
int get_max();
int del_max();
void sink(int i);
void swim(int i);
void swap(int i, int j);
void show_items(int *arr, int n);

/* Global Variables */
int *pq; 
int N; 
int capacity;

/* Test Client */
int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: ./MaxPQ <test file.txt>\n");
        return 1;
    }
    char test[1024];
    getcwd(test, 1024);
    strcat(test, "/Tests/");
    strcat(test, argv[1]);
    char buffer[255];
    FILE *file = fopen(test, "r");
    fgets(buffer, 255, file);
    int capacity = atoi(buffer);
    N = 0;
    pq = malloc(sizeof(int) * (capacity + 1));
    while (fgets(buffer, 255, file) != NULL)
    {
        insert(atoi(buffer));
    }
    int arr[capacity];
    for (int n = 0; n < capacity; n++)
    {
        arr[n] = del_max();
    }
    show_items(arr, capacity);
    free(pq);
    fclose(file);
    return 0;
}

/* Priority Queue Functions */
int insert(int item)
{
    if (N == capacity - 1)
    {
        printf("pq at max capacity: %i\n", capacity);
        return 1;
    }
    pq[++N] = item;
    swim(N);
    return 0;
}

int is_empty()
{
    if (N == 0) { return 0; }
    else { return 1; }
}

int get_max()
{
    return pq[1];
}

int del_max()
{
    int max = get_max();
    swap(1, N--);
    sink(1);
    return max;
}

void sink(int i)
{
    while (2*i <= N)
    {
        int j = 2*i; // left child
        if (j < N && pq[j] < pq[j + 1]) { j++; } // find out which child is greater
        if (pq[i] < pq[j])  // if parent less than child
        {
            swap(i, j);  // swap parent and child
            i = j; // sink down tree
        }
        else { break; }
    }
}

void swim(int i)
{
    while (i > 1 && pq[i/2] < pq[i])  // while parent is less than child
    {
        swap(i, i/2);  // swap parent and child
        i = i/2; // swim up tree
    }
}

void swap(int i, int j)
{
    int temp = pq[i];
    pq[i] = pq[j];
    pq[j] = temp;
}

void show_items(int *arr, int n)
{
    printf("[");
    for (int i = 0; i < n; i++)
    {
        if (i == n - 1)
        {
            printf("%i]\n", arr[i]);
            break;
        }
        printf("%i, ", arr[i]);
    }
}

File: DataStructures/StacksAndQueues/Queues/MaxPQ.java
package DataStructures.StacksAndQueues.Queues;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

/* A Priority Queue using a binary heap */

public class MaxPQ<Key extends Comparable<Key>> {

    private Key[] pq;
    private int N;

    public MaxPQ(int capacity) {
        pq = (Key[]) new Comparable[capacity + 1];
        N = 0;
    }

    public boolean isEmpty() {
        return N == 0;
    }

    public void insert(Key k) {
        pq[++N] = k;
        swim(N);
    }

    public Key getMax() {
        return pq[1];
    }

    /**
     * Returns and removes max from priority queue
     */
    public Key delMax() {
        Key max = getMax();
        swap(1, N--);
        sink(1);
        pq[N + 1] = null; // prevent loitering
        return max;
    }

    /**
     * Go up tree and exchage k with parent if k > parent (k/2).
     */
    private void swim(int k) {
        // while k not at root of heap and k > its parent (k/2)
        while (k > 1 && lessThan(k/2, k)) {
            swap(k, k/2);   // swap parent and child
            k = k/2;        // move up the tree
        }
    }

    /**
     * Go down tree and exchange k with larger child. 
     * Children of k are 2k and 2k + 1.
     */
    private void sink(int k) {
        while (2*k <= N) {
            int j = 2*k;    // left child
            if (j < N && lessThan(j, j + 1)) j++; // case where r child is greater
            if (!lessThan(k, j)) break;  // k is larger than both children
            swap(k, j);  // swap k with largest child
            k = j;  // go down tree
        }
    }

    private boolean lessThan(int i, int j) {
        return pq[i].compareTo(pq[j]) < 0;
    }

    private void swap(int i, int j) {
        Key temp = pq[i];
        pq[i] = pq[j];
        pq[j] = temp;
    }

    /* Test Client */
    public static void main(String[] args) throws IOException {
        String test = args[0];
        File testFile = new File(System.getProperty("user.dir") + "/Tests/" + test);
        BufferedReader br = new BufferedReader(new FileReader(testFile));
        Integer size = Integer.parseInt(br.readLine());
        MaxPQ<Integer> pq = new MaxPQ<>(size);
        String next = br.readLine();
        while (next != null) {
            if (next.equals("pop")) {
                System.out.println("Max Item: " + String.valueOf(pq.delMax()));
            } else {
                pq.insert(Integer.parseInt(next));
            }
            next = br.readLine();
        }
        while (!pq.isEmpty()) {
            System.out.println(pq.delMax());
        }
        br.close();
    }
}